qcodes/dataset/dependencies.py

```diff
@@ -1,7 +1,491 @@
+                self._raise_from(ValueError, 'Invalid standalones',
+                                 error, message)
+
+        self._remove_duplicates(dependencies)
+        self._remove_duplicates(inferences)
+
+        self.dependencies: ParamSpecTree = dependencies
+        self.inferences: ParamSpecTree = inferences
+        self.standalones: FrozenSet[ParamSpecBase] = frozenset(standalones)
+
+        # The object is now complete, but for convenience, we form some more
+        # private attributes for easy look-up
+
+        self._id_to_paramspec: Dict[str, ParamSpecBase] = {}
+        for tree in (self.dependencies, self.inferences):
+            for ps, ps_tup in tree.items():
+                self._id_to_paramspec.update({self._id(ps): ps})
+                self._id_to_paramspec.update({self._id(pst): pst
+                                              for pst in ps_tup})
+        for ps in self.standalones:
+            self._id_to_paramspec.update({self._id(ps): ps})
+        self._paramspec_to_id = {v: k for k, v in self._id_to_paramspec.items()}
+
+        self._dependencies_inv: ParamSpecTree = {}
+        indeps: List[ParamSpecBase] = []
+        for indep_tup in self.dependencies.values():
+            for indep in indep_tup:
+                indeps.append(indep)
+        for indep in set(indeps):
+            deps = tuple(ps for ps in self.dependencies
+                         if indep in self.dependencies[ps])
+            self._dependencies_inv[indep] = deps
+
+        self._inferences_inv: ParamSpecTree = {}
+        bases: List[ParamSpecBase] = []
+        for basis_tup in self.inferences.values():
+            for basis in basis_tup:
+                bases.append(basis)
+        for basis in set(bases):
+            inffs = tuple(ps for ps in self.inferences
+                          if basis in self.inferences[ps])
+            self._inferences_inv[basis] = inffs
+
+    @staticmethod
+    def _remove_duplicates(tree: ParamSpecTree) -> None:
+        """
+        Helper function to remove duplicate entries from a ParamSpecTree. Will
+        turn {A: (B, B, C)} into {A: (B, C)}
+        """
+        for ps, tup in tree.items():
+            specs: List[ParamSpecBase] = []
+            for p in tup:
+                if p not in specs:
+                    specs.append(p)
+            tree[ps] = tuple(specs)
+
+    @staticmethod
+    def _id(ps: ParamSpecBase) -> str:
+        """
+        Function to generate an ID from/for a ParamSpecBase
+        """
+        return ps.name
+
+    @staticmethod
+    def _raise_from(new_error: Type[Exception], new_mssg: str,
+                    old_error: Type[Exception], old_mssg: str) -> None:
+        """
+        Helper function to raise an error with a cause in a way that our test
+        suite can digest
+        """
+        try:
+            raise old_error(old_mssg)
+        except old_error as e:
+            raise new_error(new_mssg) from e
```

Comment:
I don't disagree, but given how much effort it seems to be to figure out how to assert that Exception1 is raised from Exception2 and how relatively small the pollution is, I'm not going to touch this. If anyone else can make `raise A from B` be detected correctly in the tests, I'll be very happy.

Reply:
So William and I did some investigation. Currently we are using the file location part of the execinfo chain to check for the original error (the [1] element of a chain member) this does not exist for the original error when you do `raise foo() from bar()`. However the exception can also be found in the traceback part of the chain ([0]th element) Here one however has to be careful as the traceback of the outer exception will contain both errors. 

What I think might work is to just check the traceback of the [0]th element of the chain which will only contain the original error. This will change the semantics of `error_caused_by` from matching any error in the chain to only the innermost error. But is that not actually what we want?

Reply:
Thanks Jens for looking into it. I'd prefer to have a new `error_caused_by_innermost` (with a better name) function that we can use in our tests. So basically, the instructions will be: use `error_caused_by` to look for the exception through the entire chain, and use `error_caused_by_innermost` to check the innermost exception that sparked the whole shebang.

Is this reasonable?